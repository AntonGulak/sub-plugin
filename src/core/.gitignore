node_modules
.env
cache
artifacts
typechain
.idea
bytecode
hashCalc.js