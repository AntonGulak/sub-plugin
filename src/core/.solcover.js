const fs = require('fs');

const testContracts = fs.readdirSync("./contracts/test")


module.exports = {
    skipFiles: testContracts.map((x) => "test/" + x),
    testfiles: "test/*.ts"
  };