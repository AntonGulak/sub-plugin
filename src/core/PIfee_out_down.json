{"tickDs":[1,2],"results":["60","70"]}