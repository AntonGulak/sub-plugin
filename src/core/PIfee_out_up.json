{"tickDs":[500,501,502,503,504,505,506,507,508,509,510,511,512,513,514,515,516,517,518,519,520,521,522,523,524,525,526,527,528,529,530,531,532,533,534,535,536,537,538,539,540,541,542,543,544,545,546,547,548,549,550,551,552,553,554,555,556,557,558,559,560,561,562,563,564,565,566,567,568,569,570,571,572,573,574,575,576,577,578,579,580,581,582,583,584,585,586,587,588,589,590,591,592,593,594,595,596,597,598,599,600,601,602,603,604,605,606,607,608,609,610,611,612,613,614,615,616,617,618,619,620,621,622,623,624,625,626,627,628,629,630,631,632,633,634,635,636,637,638,639,640,641,642,643,644,645,646,647,648,649,650,651,652,653,654,655,656,657,658,659,660,661,662,663,664,665,666,667,668,669,670,671,672,673,674,675,676,677,678,679,680,681,682,683,684,685,686,687,688,689,690,691,692,693,694,695,696,697,698,699],"results":["7929","7935","7941","7947","7953","7959","7965","7970","7976","7982","7988","7994","8000","8006","8012","8017","8023","8029","8035","8041","8047","8053","8059","8065","8070","8076","8082","8088","8094","8100","8106","8112","8117","8123","8129","8135","8141","8147","8153","8159","8165","8170","8176","8182","8188","8194","8200","8206","8212","8218","8223","8229","8235","8241","8247","8253","8259","8265","8271","8276","8282","8288","8294","8300","8306","8312","8318","8324","8329","8335","8341","8347","8353","8359","8365","8371","8376","8382","8388","8394","8400","8406","8412","8418","8424","8430","8435","8441","8447","8453","8459","8465","8471","8477","8483","8488","8494","8500","8506","8512","8518","8524","8530","8536","8541","8547","8553","8559","8565","8571","8577","8583","8589","8594","8600","8606","8612","8618","8624","8630","8636","8642","8647","8653","8659","8665","8671","8677","8683","8689","8695","8701","8706","8712","8718","8724","8730","8736","8742","8748","8754","8759","8765","8771","8777","8783","8789","8795","8801","8807","8813","8818","8824","8830","8836","8842","8848","8854","8860","8866","8872","8877","8883","8889","8895","8901","8907","8913","8919","8925","8930","8936","8942","8948","8954","8960","8966","8972","8978","8984","8989","8995","9001","9007","9013","9019","9025","9031","9037","9043","9048","9054","9060","9066","9072","9078","9084","9090","9096","9102"]}