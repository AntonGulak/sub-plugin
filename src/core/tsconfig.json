{
  "compilerOptions": {
    "target": "es2020",
    "module": "commonjs",
    "strict": true,
    "esModuleInterop": true,
    "outDir": "dist",
    "forceConsistentCasingInFileNames": true,
    "skipLibCheck": true,
    "moduleResolution": "node",
    "incremental": true,
    "typeRoots": ["./typechain", "./node_modules/@types"],
    "types": ["@nomiclabs/hardhat-ethers"]
  },
  "include": ["./test/**/*"],
  "files": ["./hardhat.config.ts"]
}
