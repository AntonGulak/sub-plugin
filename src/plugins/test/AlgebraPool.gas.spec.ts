import { ethers } from 'hardhat';
import { Wallet } from 'ethers';
import { loadFixture } from '@nomicfoundation/hardhat-network-helpers';
import { MockTimeAlgebraPool, TestERC20, TestAlgebraCallee } from '../../core/typechain'
import { MockFactory, MockPool, MockTimeDataStorageOperator, MockTimeDSFactory, DataStorageFactory } from "../typechain";
import { expect } from './shared/expect'

import { algebraPoolDeployerMockFixture } from './shared/externalFixtures'

import snapshotGasCost from './shared/snapshotGasCost'

import {
  expandTo18Decimals,
  FeeAmount,
  getMinTick,
  encodePriceSqrt,
  TICK_SPACINGS,
  createPoolFunctions,
  SwapFunction,
  MintFunction,
  getMaxTick,
  MaxUint128,
  SwapToPriceFunction,
  MAX_SQRT_RATIO,
  MIN_SQRT_RATIO,
} from '../../core/test/shared/utilities'

describe('AlgebraPool gas tests [ @skip-on-coverage ]', () => {
  let wallet: Wallet, other: Wallet;

  let token0: TestERC20;
  let token1: TestERC20;
  let swapTarget: TestAlgebraCallee;

  before('create fixture loader', async () => {
    [wallet, other] = await (ethers as any).getSigners();
  });

  for (const communityFee of [0, 60]) {
    describe(communityFee > 0 ? 'fee is on' : 'fee is off', () => {
      const startingPrice = encodePriceSqrt(100001, 100000);
      const startingTick = 0
      const tickSpacing = TICK_SPACINGS[FeeAmount.MEDIUM];
      const minTick = getMinTick(tickSpacing);
      const maxTick = getMaxTick(tickSpacing);

      const gasTestFixture = async () => {
        const fix = await algebraPoolDeployerMockFixture()
        const pool = await fix.createPool()

        const mockPluginFactoryFactory = await ethers.getContractFactory('MockTimeDSFactory')
        const mockPluginFactory = (await mockPluginFactoryFactory.deploy(fix.factory)) as any as MockTimeDSFactory
      
        await mockPluginFactory.createPlugin(pool);
        const pluginAddress = await mockPluginFactory.pluginsForPools(pool);
      
        const mockDSOperatorFactory = await ethers.getContractFactory('MockTimeDataStorageOperator')
        const plugin = mockDSOperatorFactory.attach(pluginAddress) as any as MockTimeDataStorageOperator;

        await pool.setPlugin(plugin);

        const advanceTime = async (secs: any) => {
          await pool.advanceTime(secs);
          await plugin.advanceTime(secs);
        }

        const { swapExact0For1, swapExact1For0, swapToHigherPrice, mint, swapToLowerPrice} = createPoolFunctions({
          swapTarget: fix.swapTargetCallee,
          token0: fix.token0,
          token1: fix.token1,
          pool,
        })

        token0 = fix.token0
        token1 = fix.token1
        swapTarget = fix.swapTargetCallee;

        await pool.initialize(encodePriceSqrt(1, 1))
        if (communityFee != 0)
          await pool.setCommunityFee(communityFee)
        await advanceTime(1)
        await mint(wallet.address, minTick, maxTick, expandTo18Decimals(2))
        await swapExact0For1(expandTo18Decimals(1), wallet.address)
        await advanceTime(1)
        await swapToHigherPrice(startingPrice, wallet.address)
        await advanceTime(1)
        expect((await pool.globalState()).tick).to.eq(startingTick)
        expect((await pool.globalState()).price).to.eq(startingPrice)

        return { advanceTime, pool, plugin, swapExact0For1, swapExact1For0, mint, swapToHigherPrice, swapToLowerPrice}
      }

      let swapExact0For1: SwapFunction;
      let swapExact1For0: SwapFunction;
      let swapToHigherPrice: SwapToPriceFunction;
      let swapToLowerPrice: SwapToPriceFunction;
      let pool: MockTimeAlgebraPool;
      let plugin: MockTimeDataStorageOperator;
      let mint: MintFunction
      let advanceTime: any;

      beforeEach('load the fixture', async () => {
        ;({ advanceTime, swapExact0For1, swapExact1For0, pool, plugin, mint, swapToHigherPrice, swapToLowerPrice} = await loadFixture(gasTestFixture))
      })

      describe('#swapExact1For0', () => {
        it('first swap in block with no tick movement', async () => {
          await snapshotGasCost(swapExact1For0(2000, wallet.address));
          expect((await pool.globalState()).price).to.not.eq(startingPrice);
          expect((await pool.globalState()).tick).to.eq(startingTick);
        });

        it('first swap in block moves tick, no initialized crossings', async () => {
          await snapshotGasCost(swapExact1For0(expandTo18Decimals(1) / 10000n, wallet.address));
          expect((await pool.globalState()).tick).to.eq(startingTick + 1);
        });

        it('second swap in block with no tick movement', async () => {
          await swapExact1For0(expandTo18Decimals(1) / 10000n, wallet.address);
          expect((await pool.globalState()).tick).to.eq(startingTick + 1);
          await snapshotGasCost(swapExact1For0(2000, wallet.address));
          expect((await pool.globalState()).tick).to.eq(startingTick + 1);
        });
      })

      describe('#swapExact0For1', () => {
        it('first swap in block with no tick movement', async () => {
          await snapshotGasCost(swapExact0For1(2000, wallet.address));
          expect((await pool.globalState()).price).to.not.eq(startingPrice);
          expect((await pool.globalState()).tick).to.eq(startingTick);
        });

        it('first swap in block moves tick, no initialized crossings', async () => {
          await snapshotGasCost(swapExact0For1(expandTo18Decimals(1) / 10000n, wallet.address));
          expect((await pool.globalState()).tick).to.eq(startingTick - 1);
        });

        it('second swap in block with no tick movement', async () => {
          await swapExact0For1(expandTo18Decimals(1) / 10000n, wallet.address);
          expect((await pool.globalState()).tick).to.eq(startingTick - 1);
          await snapshotGasCost(swapExact0For1(2000, wallet.address));
          expect((await pool.globalState()).tick).to.eq(startingTick - 1);
        });

        it('second swap in block moves tick, no initialized crossings', async () => {
          await swapExact0For1(1000, wallet.address);
          expect((await pool.globalState()).tick).to.eq(startingTick);
          await snapshotGasCost(swapExact0For1(expandTo18Decimals(1) / 10000n, wallet.address));
          expect((await pool.globalState()).tick).to.eq(startingTick - 1);
        });

        it('first swap in block, large swap, no initialized crossings', async () => {
          await snapshotGasCost(swapExact0For1(expandTo18Decimals(10), wallet.address));
          expect((await pool.globalState()).tick).to.eq(-35829);
        });

        it('first swap in block, large swap crossing several initialized ticks', async () => {
          await mint(wallet.address, startingTick - 3 * tickSpacing, startingTick - tickSpacing, expandTo18Decimals(1));
          await mint(
            wallet.address,
            startingTick - 4 * tickSpacing,
            startingTick - 2 * tickSpacing,
            expandTo18Decimals(1)
          );
          expect((await pool.globalState()).tick).to.eq(startingTick);
          await snapshotGasCost(swapExact0For1(expandTo18Decimals(1), wallet.address));
          expect((await pool.globalState()).tick).to.be.lt(startingTick - 4 * tickSpacing); // we crossed the last tick
        });

        it('several large swaps with pauses', async () => {
          await mint(wallet.address, startingTick - 3 * tickSpacing, startingTick - tickSpacing, expandTo18Decimals(1));
          await mint(
            wallet.address,
            startingTick - 4 * tickSpacing,
            startingTick - 2 * tickSpacing,
            expandTo18Decimals(1)
          );
          expect((await pool.globalState()).tick).to.eq(startingTick);
          await swapExact0For1(expandTo18Decimals(1) / 10000n, wallet.address);
          await pool.advanceTime(60);
          await swapExact0For1(expandTo18Decimals(1) / 10000n, wallet.address);
          await pool.advanceTime(60 * 60);
          await swapExact1For0(expandTo18Decimals(1) / 10000n, wallet.address);
          await pool.advanceTime(60 * 60);
          await swapExact0For1(expandTo18Decimals(1) / 10000n, wallet.address);
          await pool.advanceTime(5 * 60 * 60);
          await swapExact1For0(expandTo18Decimals(1) / 10000n, wallet.address);
          await pool.advanceTime(19 * 60 * 60);
          await swapExact1For0(expandTo18Decimals(1) / 10000n, wallet.address);
          await pool.advanceTime(60);
          await snapshotGasCost(swapExact0For1(expandTo18Decimals(1), wallet.address));
          expect((await pool.globalState()).tick).to.be.lt(startingTick - 4 * tickSpacing); // we crossed the last tick
        });

        it('small swap after several large swaps with pauses', async () => {
          await mint(wallet.address, startingTick - 3 * tickSpacing, startingTick - tickSpacing, expandTo18Decimals(1));
          await mint(
            wallet.address,
            startingTick - 4 * tickSpacing,
            startingTick - 2 * tickSpacing,
            expandTo18Decimals(1)
          );
          expect((await pool.globalState()).tick).to.eq(startingTick);
          await swapExact0For1(expandTo18Decimals(1) / 10000n, wallet.address);
          await pool.advanceTime(60);
          await swapExact0For1(expandTo18Decimals(1) / 10000n, wallet.address);
          await pool.advanceTime(60 * 60);
          await swapExact1For0(expandTo18Decimals(1) / 10000n, wallet.address);
          await pool.advanceTime(60 * 60);
          await swapExact0For1(expandTo18Decimals(1) / 10000n, wallet.address);
          await pool.advanceTime(5 * 60 * 60);
          await swapExact1For0(expandTo18Decimals(1) / 10000n, wallet.address);
          await pool.advanceTime(19 * 60 * 60);
          await swapExact0For1(expandTo18Decimals(1) / 10000n, wallet.address);
          await pool.advanceTime(60);
          await snapshotGasCost(swapExact0For1(1000, wallet.address));
          expect((await pool.globalState()).tick).to.be.lt(0);
        });

        it('first swap in block, large swap crossing a single initialized tick', async () => {
          await mint(wallet.address, minTick, startingTick - 2 * tickSpacing, expandTo18Decimals(1));
          await snapshotGasCost(swapExact0For1(expandTo18Decimals(1), wallet.address));
          expect((await pool.globalState()).tick).to.be.lt(startingTick - 2 * tickSpacing); // we crossed the last tick
        });

        it('second swap in block, large swap crossing several initialized ticks', async () => {
          await mint(wallet.address, startingTick - 3 * tickSpacing, startingTick - tickSpacing, expandTo18Decimals(1));
          await mint(
            wallet.address,
            startingTick - 4 * tickSpacing,
            startingTick - 2 * tickSpacing,
            expandTo18Decimals(1)
          );
          await swapExact0For1(expandTo18Decimals(1) / 10000n, wallet.address);
          await snapshotGasCost(swapExact0For1(expandTo18Decimals(1), wallet.address));
          expect((await pool.globalState()).tick).to.be.lt(startingTick - 4 * tickSpacing);
        });

        it('second swap in block, large swap crossing a single initialized tick', async () => {
          await mint(wallet.address, minTick, startingTick - 2 * tickSpacing, expandTo18Decimals(1));
          await swapExact0For1(expandTo18Decimals(1) / 10000n, wallet.address);
          expect((await pool.globalState()).tick).to.be.gt(startingTick - 2 * tickSpacing); // we didn't cross the initialized tick
          await snapshotGasCost(swapExact0For1(expandTo18Decimals(1), wallet.address));
          expect((await pool.globalState()).tick).to.be.lt(startingTick - 2 * tickSpacing); // we crossed the last tick
        });

        it('large swap crossing several initialized ticks after some time passes', async () => {
          await mint(wallet.address, startingTick - 3 * tickSpacing, startingTick - tickSpacing, expandTo18Decimals(1));
          await mint(
            wallet.address,
            startingTick - 4 * tickSpacing,
            startingTick - 2 * tickSpacing,
            expandTo18Decimals(1)
          );
          await swapExact0For1(2, wallet.address);
          await pool.advanceTime(1);
          await snapshotGasCost(swapExact0For1(expandTo18Decimals(1), wallet.address));
          expect((await pool.globalState()).tick).to.be.lt(startingTick - 4 * tickSpacing);
        });

        it('large swap crossing several initialized ticks second time after some time passes', async () => {
          await mint(wallet.address, startingTick - 3 * tickSpacing, startingTick - tickSpacing, expandTo18Decimals(1));
          await mint(
            wallet.address,
            startingTick - 4 * tickSpacing,
            startingTick - 2 * tickSpacing,
            expandTo18Decimals(1)
          );
          await swapExact0For1(expandTo18Decimals(1), wallet.address);
          await swapToHigherPrice(startingPrice, wallet.address);
          await pool.advanceTime(1);
          await snapshotGasCost(swapExact0For1(expandTo18Decimals(1), wallet.address));
          expect((await pool.globalState()).tick).to.be.lt(tickSpacing * -4);
        });

        describe('Filled DataStorage', function() {
          this.timeout(600_000)

          const filledStorageFixture = async () => {
            await mint(wallet.address, startingTick - 3 * tickSpacing, startingTick - tickSpacing, expandTo18Decimals(1))
            await mint(
              wallet.address,
              startingTick - 4 * tickSpacing,
              startingTick - 2 * tickSpacing,
              expandTo18Decimals(1)
            )
            expect((await pool.globalState()).tick).to.eq(startingTick)

            const BATCH_SIZE = 300;
            let summaryTimeDelta = 0;
            for (let i = 0; i < 1500; i += BATCH_SIZE) {
              
              const batch = [];
              for (let j = 0; j < BATCH_SIZE; j++) {
                const timeDelta = (i + j) % 2 == 0 ? 60 : 90;
                summaryTimeDelta += timeDelta;
                batch.push({
                  advanceTimeBy: timeDelta,
                  tick: startingTick + i - j
                })
              }
              await plugin.batchUpdate(batch)
            }
            await pool.advanceTime(summaryTimeDelta);
          }

          beforeEach('load inner fixture', async() => {
            await loadFixture(filledStorageFixture);
          })

          it('small swap with filled dataStorage', async () => {
            await advanceTime(15);
            await snapshotGasCost(swapExact0For1(1000, wallet.address))
          })

          it('small swap with filled dataStorage after 4h', async () => {
            await advanceTime(4 * 60 * 60);
            await snapshotGasCost(swapExact0For1(1000, wallet.address))
          })

          it('small swap with filled dataStorage after 8h', async () => {
            await advanceTime(8 * 60 * 60);
            await snapshotGasCost(swapExact0For1(1000, wallet.address))
          })

          it('large swap crossing several initialized ticks', async () => {
            await mint(wallet.address, startingTick - 3 * tickSpacing, startingTick - tickSpacing, expandTo18Decimals(1))
            await mint(
              wallet.address,
              startingTick - 4 * tickSpacing,
              startingTick - 2 * tickSpacing,
              expandTo18Decimals(1)
            )
            await swapExact0For1(2, wallet.address)
            await advanceTime(15);
            await snapshotGasCost(swapExact0For1(expandTo18Decimals(1), wallet.address))
            expect((await pool.globalState()).tick).to.be.lt(startingTick - 4 * tickSpacing)
          })
        })
      })

      describe('#mint', () => {
        for (const { description, bottomTick, topTick } of [
          {
            description: 'around current price',
            bottomTick: startingTick - tickSpacing,
            topTick: startingTick + tickSpacing,
          },
          {
            description: 'below current price',
            bottomTick: startingTick - 2 * tickSpacing,
            topTick: startingTick - tickSpacing,
          },
          {
            description: 'above current price',
            bottomTick: startingTick + tickSpacing,
            topTick: startingTick + 2 * tickSpacing,
          },
        ]) {
          describe(description, () => {
            it('new position mint first in range', async () => {
              await snapshotGasCost(mint(wallet.address, bottomTick, topTick, expandTo18Decimals(1)));
            });
            it('add to position existing', async () => {
              await mint(wallet.address, bottomTick, topTick, expandTo18Decimals(1));
              await snapshotGasCost(mint(wallet.address, bottomTick, topTick, expandTo18Decimals(1)));
            });
            it('second position in same range', async () => {
              await mint(wallet.address, bottomTick, topTick, expandTo18Decimals(1));
              await snapshotGasCost(mint(other.address, bottomTick, topTick, expandTo18Decimals(1)));
            });
            it('add to position after some time passes', async () => {
              await mint(wallet.address, bottomTick, topTick, expandTo18Decimals(1));
              await pool.advanceTime(1);
              await snapshotGasCost(mint(wallet.address, bottomTick, topTick, expandTo18Decimals(1)));
            });
          });
        }
      });

      describe('#burn', () => {
        for (const { description, bottomTick, topTick } of [
          {
            description: 'around current price',
            bottomTick: startingTick - tickSpacing,
            topTick: startingTick + tickSpacing,
          },
          {
            description: 'below current price',
            bottomTick: startingTick - 2 * tickSpacing,
            topTick: startingTick - tickSpacing,
          },
          {
            description: 'above current price',
            bottomTick: startingTick + tickSpacing,
            topTick: startingTick + 2 * tickSpacing,
          },
        ]) {
          describe(description, () => {
            const liquidityAmount = expandTo18Decimals(1);
            beforeEach('mint a position', async () => {
              await mint(wallet.address, bottomTick, topTick, liquidityAmount);
            });

            it('burn when only position using ticks', async () => {
              await snapshotGasCost(pool.burn(bottomTick, topTick, expandTo18Decimals(1), '0x'));
            });
            it('partial position burn', async () => {
              await snapshotGasCost(pool.burn(bottomTick, topTick, expandTo18Decimals(1) / 2n, '0x'));
            });
            it('entire position burn but other positions are using the ticks', async () => {
              await mint(other.address, bottomTick, topTick, expandTo18Decimals(1));
              await snapshotGasCost(pool.burn(bottomTick, topTick, expandTo18Decimals(1), '0x'));
            });
            it('burn entire position after some time passes', async () => {
              await pool.advanceTime(1);
              await snapshotGasCost(pool.burn(bottomTick, topTick, expandTo18Decimals(1), '0x'));
            });
          });
        }
      });

      describe('#poke', () => {
        const bottomTick = startingTick - tickSpacing;
        const topTick = startingTick + tickSpacing;

        it('best case', async () => {
          await mint(wallet.address, bottomTick, topTick, expandTo18Decimals(1));
          await swapExact0For1(expandTo18Decimals(1)/ 100n, wallet.address);
          await pool.burn(bottomTick, topTick, 0, '0x');
          await swapExact0For1(expandTo18Decimals(1)/ 100n, wallet.address);
          await snapshotGasCost(pool.burn(bottomTick, topTick, 0, '0x'));
        });
      });

      describe('#collect', () => {
        const bottomTick = startingTick - tickSpacing;
        const topTick = startingTick + tickSpacing;

        it('close to worst case', async () => {
          await mint(wallet.address, bottomTick, topTick, expandTo18Decimals(1));
          await swapExact0For1(expandTo18Decimals(1)/ 100n, wallet.address);
          await pool.burn(bottomTick, topTick, 0, '0x'); // poke to accumulate fees
          await snapshotGasCost(pool.collect(wallet.address, bottomTick, topTick, MaxUint128, MaxUint128));
        });

        it('close to worst case, two tokens', async () => {
          await mint(wallet.address, bottomTick, topTick, expandTo18Decimals(1));
          await swapExact0For1(expandTo18Decimals(1)/ 100n, wallet.address);
          await swapExact1For0(expandTo18Decimals(1)/ 100n, wallet.address);
          await pool.burn(bottomTick, topTick, 0, '0x'); // poke to accumulate fees
          await snapshotGasCost(pool.collect(wallet.address, bottomTick, topTick, MaxUint128, MaxUint128));
        });
      });
    });
  }
});
