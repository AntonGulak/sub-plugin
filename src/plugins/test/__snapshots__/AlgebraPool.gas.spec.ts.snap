// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AlgebraPool gas tests [ @skip-on-coverage ] Filled VolatilityOracle swaps dynamic fee large swap crossing several initialized ticks 1`] = `229253`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] Filled VolatilityOracle swaps dynamic fee small swap with filled volatilityOracle 1`] = `178713`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] Filled VolatilityOracle swaps dynamic fee small swap with filled volatilityOracle after 4h 1`] = `184682`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] Filled VolatilityOracle swaps dynamic fee small swap with filled volatilityOracle after 8h 1`] = `189640`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] Filled VolatilityOracle swaps static fee large swap crossing several initialized ticks 1`] = `226367`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] Filled VolatilityOracle swaps static fee small swap with filled volatilityOracle 1`] = `175869`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] Filled VolatilityOracle swaps static fee small swap with filled volatilityOracle after 4h 1`] = `181838`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] Filled VolatilityOracle swaps static fee small swap with filled volatilityOracle after 8h 1`] = `186796`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] Positions #burn above current price burn when only position using ticks 1`] = `154290`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] Positions #burn above current price entire position burn but other positions are using the ticks 1`] = `155171`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] Positions #burn above current price partial position burn 1`] = `159971`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] Positions #burn around current price burn when only position using ticks 1`] = `162436`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] Positions #burn around current price entire position burn but other positions are using the ticks 1`] = `159585`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] Positions #burn around current price partial position burn 1`] = `164385`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] Positions #burn below current price burn when only position using ticks 1`] = `161975`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] Positions #burn below current price entire position burn but other positions are using the ticks 1`] = `155815`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] Positions #burn below current price partial position burn 1`] = `160615`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] Positions #collect close to worst case 1`] = `52532`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] Positions #collect close to worst case, two tokens 1`] = `70251`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] Positions #mint above current price add to position existing 1`] = `172749`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] Positions #mint above current price new position mint first in range 1`] = `326639`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] Positions #mint above current price second position in same range 1`] = `189849`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] Positions #mint around current price add to position existing 1`] = `197826`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] Positions #mint around current price new position mint first in range 1`] = `404749`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] Positions #mint around current price second position in same range 1`] = `214926`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] Positions #mint below current price add to position existing 1`] = `173348`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] Positions #mint below current price new position mint first in range 1`] = `401003`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] Positions #mint below current price second position in same range 1`] = `190448`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] Positions #poke best case 1`] = `108010`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is off #swap #swapExact0For1 first swap in block moves tick, no initialized crossings 1`] = `149003`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is off #swap #swapExact0For1 first swap in block with no tick movement 1`] = `148972`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is off #swap #swapExact0For1 first swap in block with no tick movement, static fee 1`] = `149669`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is off #swap #swapExact0For1 first swap in block, large swap crossing a single initialized tick 1`] = `165631`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is off #swap #swapExact0For1 first swap in block, large swap crossing several initialized ticks 1`] = `199358`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is off #swap #swapExact0For1 first swap in block, large swap, no initialized crossings 1`] = `149070`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is off #swap #swapExact0For1 large swap crossing several initialized ticks after some time passes 1`] = `199358`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is off #swap #swapExact0For1 large swap crossing several initialized ticks second time after some time passes 1`] = `218558`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is off #swap #swapExact0For1 second swap in block moves tick, no initialized crossings 1`] = `113456`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is off #swap #swapExact0For1 second swap in block with no tick movement 1`] = `113420`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is off #swap #swapExact0For1 second swap in block, large swap crossing a single initialized tick 1`] = `129269`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is off #swap #swapExact0For1 second swap in block, large swap crossing several initialized ticks 1`] = `163817`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is off #swap #swapExact0For1 several large swaps with pauses 1`] = `227874`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is off #swap #swapExact0For1 small swap after several large swaps with pauses 1`] = `158157`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is off #swap #swapExact1For0 first swap in block moves tick, no initialized crossings 1`] = `149094`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is off #swap #swapExact1For0 first swap in block with no tick movement 1`] = `149042`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is off #swap #swapExact1For0 second swap in block with no tick movement 1`] = `113511`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is off #swap farming connected first swap in block moves tick, no initialized crossings 1`] = `180321`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is off #swap farming connected first swap in block with no tick movement 1`] = `180269`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is off #swap farming connected second swap in block with no tick movement 1`] = `127638`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is on #swap #swapExact0For1 first swap in block moves tick, no initialized crossings 1`] = `156901`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is on #swap #swapExact0For1 first swap in block with no tick movement 1`] = `156870`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is on #swap #swapExact0For1 first swap in block with no tick movement, static fee 1`] = `149906`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is on #swap #swapExact0For1 first swap in block, large swap crossing a single initialized tick 1`] = `173766`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is on #swap #swapExact0For1 first swap in block, large swap crossing several initialized ticks 1`] = `208204`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is on #swap #swapExact0For1 first swap in block, large swap, no initialized crossings 1`] = `156968`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is on #swap #swapExact0For1 large swap crossing several initialized ticks after some time passes 1`] = `208204`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is on #swap #swapExact0For1 large swap crossing several initialized ticks second time after some time passes 1`] = `227404`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is on #swap #swapExact0For1 second swap in block moves tick, no initialized crossings 1`] = `121354`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is on #swap #swapExact0For1 second swap in block with no tick movement 1`] = `121318`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is on #swap #swapExact0For1 second swap in block, large swap crossing a single initialized tick 1`] = `137404`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is on #swap #swapExact0For1 second swap in block, large swap crossing several initialized ticks 1`] = `172663`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is on #swap #swapExact0For1 several large swaps with pauses 1`] = `236720`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is on #swap #swapExact0For1 small swap after several large swaps with pauses 1`] = `158394`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is on #swap #swapExact1For0 first swap in block moves tick, no initialized crossings 1`] = `156992`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is on #swap #swapExact1For0 first swap in block with no tick movement 1`] = `156940`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is on #swap #swapExact1For0 second swap in block with no tick movement 1`] = `121409`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is on #swap farming connected first swap in block moves tick, no initialized crossings 1`] = `188219`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is on #swap farming connected first swap in block with no tick movement 1`] = `188167`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is on #swap farming connected second swap in block with no tick movement 1`] = `135536`;
